File: week-6/bstsimple/bstsimple.cpp
#include <fstream>
#include <vector>
using std::vector;

#include <iostream>
using std::cout;
using std::cin;
using std::endl;

std::ifstream fin{"bstsimple.in"};
std::ofstream fout{"bstsimple.out"};

struct Node
{
    long value;
    Node *parent;
    Node *left;
    Node *right;

    Node(long value)
    {
        this->value = value;
    }
    Node(long value, Node *left, Node *right)
    {
        this->value = value;
        this->left = left;
        this->right = right;
    }
};

class BST
{
    private:
    Node *root;

    Node* minimum(Node *x)
    {
        if (x->left == nullptr)
            return x;

        return minimum(x->left);
    }
    Node* remove1 (Node *z)
    {
        if (z->left == nullptr)
            return z->right;
        else
            return z->left;
    }
    Node* remove2 (Node *z)
    {
        auto y = minimum(z->right);
        z->right = remove(z->right, y);
        y->right = z->right;
        y->left = z->left;

        return y;
    }
    Node* search(Node *x, long value)
    {
        if (x == nullptr || value == x->value)
            return x;
        if (value < x->value)
            return search(x->left, value);
        else
            return search(x->right, value);
    }

    public:
    Node* next(Node *x)
    {
        if (x->right != nullptr)
            return minimum(x->right);

        auto y = x->parent;
        while (y != nullptr && x == y->right)
        {
            x = y;
            y = y->parent;
        }

        return y;
    }
    Node* insert(Node *x, Node *z)
    {
        if (x == nullptr)
            return z;

        if (z->value < x->value)
            x->left = insert(x->left, z);
        else
            x->right = insert(x->right, z);
    }
    Node* remove(Node *x, Node *z)
    {
        if (z->value < x->value)
        {
            x->left = remove(x->left, z);
            return x;
        }
        else if (z->value > x->value)
        {
            x->right = remove(x->right, z);
            return x;
        }
        else
        {
            if (x->left == nullptr || x->right == nullptr)
                return remove1(x);
            else
                return remove2(x);
        }
    }

    BST(long value) {
        this->root = new Node(value);
    }
};

int main()
{
    std::string cmd;
    long x;
    auto tree = new BST(5);
    tree.insert()
    // while (fin >> cmd && fin >> x)
    // {
    //     cout << cmd << endl;
    //     if (cmd == "insert")
    //     {
    //         tree->add(x);
    //         continue;
    //     }
    //     if (cmd == "delete")
    //     {
    //         tree->remove(x);
    //         continue;
    //     }
    //     if (cmd == "exists")
    //     {
    //         tree->exists(x);
    //         continue;
    //     }
    //     if (cmd == "next")
    //     {
    //         tree->next(x);
    //         continue;
    //     }
    //     if (cmd == "prev")
    //     {
    //         tree->prev(x);
    //         continue;
    //     }
    // }
    return 0;
}

File: week-6/bstsimple/bstsimple.py
class Node:
    def __init__(self, parent, value):
        self.key = value
        self.left = None
        self.right = None
        self.parent = parent

    def insert(self, value):
        if value < self.key:
            if self.left is None:
                self.left = Node(self, value)
                return self.left
            else:
                return self.left.insert(value)
        else:
            if self.right is None:
                self.right = Node(self, value)   
                return self.right
            else:
                return self.right.insert(value)

    def find(self, value):
        if value == self.key:
            return self
        elif value < self.key:
            if self.left is None:
                return None
            else:
                return self.left.find(value)
        else:
            if self.right is None:
                return None
            else:
                return self.right.find(value)
    
    def next(self):
        if self.right is not None:
            return self.right.minimum()
        current = self
        while current.parent is not None and current.parent.right is current:
            current = current.parent
        return current.parent

    def prev(self):
        if self.left is not None:
            return self.left.maximum()
        current = self
        while current.parent is not None and current.parent.left is current:
            current = current.parent
        return current.parent

    def delete(self):
        if self.left is None or self.right is None:
            if self is self.parent.left:
                self.parent.left = self.left or self.right
                if self.parent.left is not None:
                    self.parent.left.parent = self.parent
            else:
                self.parent.right = self.left or self.right
                if self.parent.right is not None:
                    self.parent.right.parent = self.parent 
            current = self.parent
            while current.key is not None:
                current = current.parent
            return self
        else:
            s = self.next()
            self.key, s.key = s.key, self.key
            return s.delete()

    def minimum(self):
        current = self
        while current.left is not None:
            current = current.left
        return current

    def maximum(self):
        current = self
        while current.right is not None:
            current = current.right
        return current

class BST:
    def __init__(self):
        self.root = None
        self.psroot = Node(None, None)

    def reroot(self):
        self.root = self.psroot.left

    def insert(self, value):
        if self.root is None:
            self.psroot.left = Node(self.psroot, value)
            self.reroot()
            return self.root
        else:
            return self.root.insert(value)

    def exists(self, value):
        if self.root is None:
            return None
        else:
            return self.root.find(value)

    def next(self, value):
        node = self.exists(value)
        if node:
            return node.next().key
        else:
            nxt = self.insert(value).next().key
            self.delete(value)
            return nxt
    
    def prev(self, value):
        node = self.exists(value)
        if node:
            return node.prev().key
        else:
            prv = self.insert(value).prev().key
            self.delete(value)
            return prv

    def delete(self, value):
        node = self.exists(value)
        deleted = self.root.delete()
        self.reroot()
        return deleted

tree = BST()

fin = open('bstsimple.in', 'r')
fout = open('bstsimple.out', 'w')
for line in fin.readlines():
    cmd, val = line.split()
    val = int(val)
    
    if cmd == 'insert':
        tree.insert(val)
        continue
    elif cmd == 'exists':
        node = tree.exists(val)
        fout.write('true\n') if node else fout.write('false\n')
        continue
    elif cmd == 'next':
        nxt = tree.next(val)
        fout.write(str(nxt).lower()+'\n')
        continue
    elif cmd == 'prev':
        prv = tree.prev(val)
        fout.write(str(prv).lower()+'\n')
        continue
    elif cmd == 'delete':
        tree.delete(val)
        continue
